File: Medium/1845. Seat Reservation Manager/1845. Seat Reservation Manager.java
import java.util.PriorityQueue;

class SeatManager {
    private int last;
    private PriorityQueue<Integer> pq;

    public SeatManager(int n) {
        last = 0;
        pq = new PriorityQueue<>();
    }

    public int reserve() {
        if (pq.isEmpty()) {
            return ++last;
        } else {
            int seat = pq.poll();
            return seat;
        }
    }

    public void unreserve(int seatNumber) {
        if (seatNumber == last) {
            last--;
        } else {
            pq.offer(seatNumber);
        }
    }
}

File: Medium/1845. Seat Reservation Manager/1845. Seat Reservation Manager.py
import heapq


class SeatManager:
    def __init__(self, n: int):
        self.last = 0
        self.pq = []

    def reserve(self) -> int:
        if not self.pq:
            self.last += 1
            return self.last
        else:
            seat = heapq.heappop(self.pq)
            return seat

    def unreserve(self, seatNumber: int) -> None:
        if seatNumber == self.last:
            self.last -= 1
        else:
            heapq.heappush(self.pq, seatNumber)
